Startx: Implement xserverArgs, improve logfile location

Atemu opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change

Closes: #80198
The Xorg log was hard-coded to ~/.xorg.log, not ideal.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

Atemu committed:
xinit: use the system xserverrc

Previously it'd try to use the one in its output path which is read-only of
course

Atemu committed:
startx: provide xserverArgs via xserverrc

Fixes #80198

Atemu committed:
xserver: make logFile configurable

It makes sense for it to be /dev/null for all the displayManagers but startx,
it needs a different logFile configuration.

Atemu committed:
startx: send Xorg log to the default location

This partially reverts bf3d3dd19b48c432dd83aa0385b47dbe84aa647b.

I don't know why we weren't getting a default logfile back then but Xorg
definitely provides one now (see `man Xorg`). It's in a much better place
aswell ($XDG_DATA_HOME)

Atemu commented:
@aszlig does this PR work for your setup which needed your change that I partially reverted?

worldofpeace commented on pkgs/servers/x11/xorg/overrides.nix:
```diff
@@ -748,8 +748,9 @@ self: super:
-      sed -i 's|^defaultserverargs="|&-logfile \"$HOME/.xorg.log\"|p' startx.cpp
+    # Link to system's xserverrc
+    fixupPhase = ''
+      ln -sf /etc/X11/xinit/xserverrc $out/etc/X11/xinit/xserverrc
```
we should never override fixupPhase, there's important things that happen in it.
Please use `postFixup`.

Atemu replied:
Thanks!

Atemu commented:
Just telling users to set some option if they want to use startx as root was not a very good solution, so I somewhat hackily fixed the log location of xorg-server which understandably requires quite a rebuild and that probably isn't ideal.

I have a solution that would not require that rebuild and would work almost as well but isn't as clean; should I go with that instead?

Atemu committed:
xorgserver: don't create log-dir during build

Xorg does this because X would crash if it couldn't write to the logfile.

Not a problem on NixOS as journald should create the dir automatically and the
logfile location is set to /dev/null for all displayManagers but startx.

Atemu committed:
xorgserver: don't override patchPhase, only append

Not sure why it was done this way. Now we can apply patches again

Atemu committed:
xorgserver: set log-dir to /var/log instead of the Nix store

It was in ${xorg.xorgserver}/var/log before which isn't writable and that would
crash X.

This required some workarounds for the startx displayManager which are no longer
needed now.

Atemu commented:
My previous patchset was clumsy, hacky and had some unecessary changes in it. What I just force-pushed should be a lot better.

Atemu commented:
I successfully ran the i3-wm test and manually confirmed that sway and i3 work.

I did not have the time to build and test every package that depends on xorgserver.

Atemu commented:
/marvin opt-in
/status needs_reviewer

doronbehar commented on nixos/modules/services/x11/display-managers/startx.nix:
```diff
@@ -39,6 +39,10 @@ in
     };
     systemd.services.display-manager.enable = false;
+
+    # Let Xorg use its default log location ($XDG_DATA_HOME/xorg/)
```
I'm a bit confused, isn't the default after all previous changes `/var/log/xorg`?

Atemu replied:
Xorg logs to `$XDG_DATA_HOME/xorg/` when run as a regular user and `/var/log/xorg` when run as root.

I enhanced the commit messages to include this info.

doronbehar commented on pkgs/servers/x11/xorg/overrides.nix:
```diff
@@ -759,9 +759,6 @@ self: super:
                          ++ lib.optionals isDarwin [ self.libX11 self.xorgproto ];
-    prePatch = ''
-      sed -i 's|^defaultserverargs="|&-logfile \"$HOME/.xorg.log\"|p' startx.cpp
-    '';
```
I think this should be part of the commit were you patched xorg.

Atemu replied:
It's one logical step: Old solution with hard-coded path in xinit -> new solution with upstream path and configurable override in NixOS that was set up in the previous commits.

doronbehar commented on pkgs/servers/x11/xorg/overrides.nix:
```diff
@@ -625,6 +625,7 @@ self: super:
       if (!isDarwin)
       then {
         outputs = [ "out" "dev" ];
+        patches = [ ./dont-create-logdir.patch ];
```
What confuses me is that this change, and where all the defaults are changed are set in different commits (the last one, and the commit after this one). If you could squash them all into once, and connect everything in the commit message, it'd be great.

Atemu replied:
The xorgserver patches don't actually change the defaults (or any behaviour in NixOS for that matter). They are patches which make workarounds we had to add in NixOS obsolete.

doronbehar commented on pkgs/servers/x11/xorg/overrides.nix:
```diff
@@ -570,7 +570,7 @@ self: super:
         if (abiCompat == null || lib.hasPrefix abiCompat version) then
           attrs_passed // {
-            buildInputs = attrs_passed.buildInputs ++ [ libdrm.dev ]; patchPhase = ''
+            buildInputs = attrs_passed.buildInputs ++ [ libdrm.dev ]; postPatch = ''
```
This commit as well may be squashed with the next one.

doronbehar commented on pkgs/servers/x11/xorg/overrides.nix:
```diff
@@ -760,6 +760,9 @@ self: super:
     '';
+    postFixup = ''
+      substituteInPlace $out/bin/startx --replace $out/etc/X11/xinit/xserverrc /etc/X11/xinit/xserverrc
+    '';
```
So there is a default for startx, which is not effected by the default set in build? - In the commits coming after this?

Atemu replied:
The default xserverrc is in the Nix store, this changes it to /etc/.

The following patches handle the logfile, not xserverrc. Technically I could've put this in a separate PR but since I was touching startx anyways, I thought I'd bundle them together.

doronbehar commented:
@timokau this PR was reviewed yet @marvin-mk2 still nags it to be reviewed.

timokau commented:
@doronbehar I think you left your review as "comment" instead of "request changes", which left the status as "awaiting review". Its debatable if this is correct or not, but at least its not a bug. My intention is to avoid having PRs stuck in "awaiting changes" where they can be forgotten, so I err'ed on the side of "awaiting review".

/status awaiting_changes

Atemu commented:
Continued in #100774
